wazuh/wazuh-kibana-app
# Add validation to the plugin settings

# Description

This PR:
- Add:
  - Plugin setting validations:
    - Create reusable methods to compose the validation of each setting
    - Frontend `Settings/Configuration` form:
      - Display the validation error
      - Highlight the setting with invalid input
      - Display the count of settings with an invalid value
      - Disable the save button when there is some invalid value in some setting changed
    - Backend:
      - `PUT /utils/configuration`
        - All settings are validated.
- Test:
  - `useForm` hook. Validate values.
  - `InputFrom` component renders the validation errors. Generic.
  - validate the value for each setting when updating from API endpoint `PUT /utils/configuration`
      
Closes #4446

# Screenshots

- Input's validation, invalid settings highlight, error count display, and disabled of the save button.
![image](https://user-images.githubusercontent.com/34042064/188813569-958c5d6d-d66c-4840-b805-0df33b283d64.png)
 
- Warning icon added to the invalid settings, reason displayed on mouseover.
![image](https://user-images.githubusercontent.com/34042064/188813794-c49c7e11-26e5-4649-927b-9dbfb54943e6.png)

# Tests

## Manual tests
**Scenario**: Set an invalid value for the setting in `Settings/Configuration` form
  **When** the user changes the input value to something invalid. This depends on each setting validation.
  **Then** an error should appear below the input form.
  **And** the setting should be red colored and highlighted as invalid.
  **And** the bottom bar should display the count settings with invalid values.
  **And** the save button should be disabled.
  
**Scenario**: Set 2 or more invalid values for the settings in `Settings/Configuration` form
  **When** the user changes the input value of 2 or more settings to invalid values. This depends on each setting validation.
  **Then** an error should appear below the input form for each setting.
  **And** the settings should be red colored and highlighted as invalid.
  **And** the bottom bar should display the count settings with invalid values.
  **And** the save button should be disabled.
  
**Scenario**: Set a valid value for the setting in `Settings/Configuration` form
  **When** the user changes the input value to a valid value. This depends on each setting validation.
  **Then** it should not appear as a validation error below the input.
  
**Scenario**: Send invalid values to update the plugin configuration through the plugin API endpoint
  **When** the user sends (using any rest client) a request to `PUT /utils/configuration` with one or more settings, where at least one of them has an invalid value.
  **Then** it should reply with a bad request displaying the validation error.
  
## Automatic tests
- `server/routes/wazuh-utils/wazuh-utils.test.ts`
- `public/components/common/form/hooks.test.tsx`
- `public/components/common/form/index.test.tsx`
- `common/plugin-settings.test.ts`

**Run all PR tests**
```shell
yarn test:jest server/routes/wazuh-utils/wazuh-utils.test.ts public/components/common/form/hooks.test.tsx public/components/common/form/index.test.tsx common/plugin-settings.test.ts
```
